File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/ContractCallDraftIntent.java
package org.hyperledger.iroha.android.client;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.model.ContractInvocation;
import org.hyperledger.iroha.android.model.JsonValue;

/**
 * Caller-trusted, local-only intent used to validate an unsigned contract-call draft.
 *
 * <p>The invocation must come from a trusted contract artifact and argument schema. Metadata is
 * the exact final transaction metadata the caller is prepared to sign. Neither value is sent to
 * Torii as proof of its own response.
 */
public final class ContractCallDraftIntent {
  private final ContractInvocation invocation;
  private final Map<String, JsonValue> metadata;

  public ContractCallDraftIntent(
      final ContractInvocation invocation, final Map<String, JsonValue> metadata) {
    this.invocation = Objects.requireNonNull(invocation, "invocation");
    final Map<String, JsonValue> snapshot = new LinkedHashMap<>();
    Objects.requireNonNull(metadata, "metadata")
        .forEach(
            (key, value) ->
                snapshot.put(
                    Objects.requireNonNull(key, "metadata key"),
                    Objects.requireNonNull(value, "metadata value")));
    this.metadata = Collections.unmodifiableMap(snapshot);
  }

  /** Returns the exact resolved invocation authorized by the caller. */
  public ContractInvocation invocation() {
    return invocation;
  }

  /** Returns the exact final transaction metadata authorized by the caller. */
  public Map<String, JsonValue> metadata() {
    return metadata;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/ContractCallDraftIntent.kt
package org.hyperledger.iroha.sdk.client

import java.util.Collections
import java.util.LinkedHashMap
import org.hyperledger.iroha.sdk.core.model.ContractInvocation
import org.hyperledger.iroha.sdk.core.model.JsonValue

/**
 * Caller-trusted contract-call intent used to verify an unsigned Torii draft.
 *
 * Construct the invocation from a trusted contract artifact and exact argument schema. The
 * metadata is the complete final transaction metadata the caller intends to sign; it is kept
 * off-wire and is never populated from response echoes.
 */
class ContractCallDraftIntent(
    @JvmField val invocation: ContractInvocation,
    metadata: Map<String, JsonValue>,
) {
    private val _metadata: Map<String, JsonValue> =
        Collections.unmodifiableMap(LinkedHashMap(metadata))

    /** Exact signature-bound transaction metadata authorized by the caller. */
    val metadata: Map<String, JsonValue> get() = _metadata

    init {
        _metadata.keys.forEach { key ->
            require(key.isNotBlank()) { "metadata key must not be blank" }
        }
    }

    override fun equals(other: Any?): Boolean =
        this === other ||
            other is ContractCallDraftIntent &&
            invocation == other.invocation &&
            _metadata == other._metadata

    override fun hashCode(): Int = 31 * invocation.hashCode() + _metadata.hashCode()
}
